### PR TITLE
AlignmentMatrixControl: keep the physical direction in RTL languages

### LINKED ISSUE #36237: __experimentalBlockAlignmentMatrixControl : the positions of the dots should not change when RTL
### Description

I tested it on RTL, and it says about the point that is "top left": "top right", even when console log the values, they are the in the opposite direction of dots:

though if you test it in cover block... the content aligns perfectly in the appropriate location as the dots... but i think that because who ever developed the cover block has used flex and flex direction change on RTL... but that is not the solution. cause now the tooltip text is different from the actual dot's location. so, I guess the direction of the dots should stay in one way for RTL and LTR... 

### Step-by-step reproduction instructions

1. console log the values  generated on change of  __experimentalBlockAlignmentMatrixControl component.
2.  go to the block editor and test your block... 
3. see what has been generate when you change alignment
4. flip html direction to RTL by changing to any RTL language.
5. see if the values u get corresponds with UI dot's location.

### Screenshots, screen recording, code snippet

image

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by t-hamano**
Fix: #36237

## What?
~~This PR fixes a problem in the RTL language where tooltip labels for the `AlignmentMatrixControl` component are not displayed correctly.~~

This PR fixes the physical position of the grid in `AlignmentMatrixControl` component, regardless of language orientation.

## Why?
In the RTL language, the grid on which the `AlignmentMatrixControl` component is written is inverted left to right.
~~However, tooltip labels do not take language orientation into account, so the left-right labels will be reversed.~~
Therefore, the physical direction indicated by the tooltip is horizontally reversed in the RTL language.

## How?
~~In the variable that defines tooltip labels,  I added `isRTL` function to separate labels.~~
Applied `direction:ltr` to a row that is part of a component.
In addition, I have added a fix to maintain RTL language behavior in the cover block, which is the only one using this component.

### Test on the storybook

- Build and open the storybook on this branch.
- In the `AlignmentMatrixControl` component, confirm that the grid is maintained when the language direction is changed and that the tooltip text also points in the correct direction.

### Test on the block editor

1. Build on `trunk` and open the post / site editor.
2. Insert a cover block and one block in it.
3. Use `AlignmentMatrixControl` to align to either left or right.
4. Use [RTL Tester plugin](https://wordpress.org/plugins/rtl-tester/) to change the language direction to RTL.
5. The blocks in the cover block are placed in reverse (The physical direction should be maintained, but is outside the scope of this issue.).
6. Ensures that the isRTL function returns `rtl` with the following code:

```
wp.hooks.addFilter( 'i18n.gettext_with_context', 'test', () => 'rtl' );
```

7. Confirm that the tooltip text shows the left and right sides reversed when the grid is focused.
8. Build on this branch and open the page.
9. Ensures that the isRTL function returns `rtl` with the following code:

```
wp.hooks.addFilter( 'i18n.gettext_with_context', 'test', () => 'rtl' );
```

10. In the cover block, confirm that the physical position of the child blocks is maintained and that the tooltip text also points in the correct direction.

### CODE REVIEW

File: packages/components/src/alignment-matrix-control/utils.js [hunk 1]
```diff
@@ -11,15 +11,15 @@ export const GRID = [
 
 // Stored as map as i18n __() only accepts strings (not variables)
 export const ALIGNMENT_LABEL = {
-	'top left': __( 'Top Left' ),
+	'top left': ! isRTL() ? __( 'Top Left' ) : __( 'Top Right' ),
```
Comment (mirka) [inline packages/components/src/alignment-matrix-control/utils.js]: Responding to [this](https://github.com/WordPress/gutenberg/issues/36237#issuecomment-1210688191):

> Note: In my environment, `isRTL` function did not respond correctly when I changed to the RTL via RTL Tester.

I think this is because the `isRTL` call here is being made in an object that is completely outside of the AlignmentMatrixControl component function. So the values becomes static when the JS bundle is loaded. For the Storybook RTL switcher to work, the `isRTL` needs to be called again when the AlignmentMatrixControl component is re-rendered. Could you try reworking it so the conditional fires again when the component is re-rendered? For example, using a `getAlignmentLabel()` function instead of accessing an [object](https://github.com/WordPress/gutenberg/blob/8a22f96709b4f6172a58e4a5de88c26576beb683/packages/components/src/alignment-matrix-control/cell.js#L18).

Hope that makes sense. It doesn't matter in actual WP usage, but it's helpful for it to be testable in Storybook if possible 🙂

**Comment by mirka** [comment]
Tagging some reviewers who've worked on the Cover block to check that this solution makes sense in that context.

**Comment by t-hamano** [comment]
I prepared a wrapper function and added processing by `isRTL` in it.
It now works fine on the storybook.


https://user-images.githubusercontent.com/54422211/184307681-b61231e1-bcd8-43c8-bf8a-046b0013872f.mp4

File: packages/components/src/alignment-matrix-control/utils.js (hunk 1 above)
Comment (mirka) [reply packages/components/src/alignment-matrix-control/utils.js]: It's working well now, thanks 🎉

**Review by mirka — COMMENTED** [review]
This is an improvement over the current behavior, so I'm fine with merging. However, I'm not sure if it's a good permanent fix.

The part that I'm worried about is that the visual labels are now mismatched with the `value` in the component state. It makes me wonder if the RTL handling should be done in a different layer.

Although the AlignmentMatixControl is currently only used in the Cover block, it would be nice to keep them from being tightly coupled. The fix proposed in this PR relies on the fact that the Cover block is converting physical directions (`left`) to logical directions (`start`):

https://github.com/WordPress/gutenberg/blob/e5154c91da00121ccc1b0fe0954799ed3a2d5ead/packages/block-library/src/cover/style.scss#L129-L132

If we imagine a consumer that does _not_ do this, and instead uses the alignment values to do physical positioning like `position: absolute; left: 0;`, the behavior is quite confusing. If the user selected the dot in the top left corner, you want the component to give you the `top left` value, regardless of LTR/RTL.

tl;dr — Because AlignmentMatrixControl is dealing with physical direction values (left, right), and there's nothing about the component that suggests it's meant for aligning text, I think it's confusing that it returns flipped values in RTL mode. The matrix dots themselves shouldn't be flipped in RTL. Any RTL flipping, if desired, should be handled by the consumer.

What do you think? In any case it's more of a follow-up and not particularly high priority, so if you think we should merge this PR for the time being, just add a changelog and I can approve 👍

**Comment by t-hamano** [comment]
Thank you for your very detailed advice 🙂
Indeed, this fix may not be theoretical.

I have considered the following approach, but is this a radical solution?

### Fix AlignmentMatrixControl to LTR

Add `direction:ltr;` below so that the grid is not inverted left to right in the RTL language.

https://github.com/WordPress/gutenberg/blob/0c1705ebc7497195dedd72a6b34dd633550d64d5/packages/components/src/alignment-matrix-control/styles/alignment-matrix-control-styles.js#L39-L43

This will maintain the direction of the grid regardless of the language direction, and labels and values should also show the physical direction.
I believe that this is the correct behavior as a stand-alone component.

## Cover block adjustment

I think that the fixed grid direction will change the existing behavior in the cover block.
Use `isRTL()` within the cover block and make adjustments to maintain existing behavior and avoid breaking existing blocks.

**Comment by mirka** [comment]
Yes, I would say that is the cleanest fix for this problem 👍 I don't know offhand if the `direction: ltr` is going to mess up the Tooltip text direction, but I agree with the general approach (i.e. don't flip the matrix, flip in Cover block).

**Comment by t-hamano** [comment]
✅ In `AlignmentMatrixControl` component, physical orientation is now maintained regardless of language.
✅ In the cover block, I have added a quick fix to maintain the previous behavior.
✅ Update changelog

In the cover block, I think the current behavior is not inherently expected.
However, the application of the physical direction to the cover block has a significant impact on existing consumers, so I would like to create another issue to discuss after this PR is merged.

**Review by mirka — APPROVED** [review]
> In the cover block, I think the current behavior is not inherently expected.

Good point. The Cover block is usually used to place text over images, and images don't get flipped in RTL. When you consider a multilingual site that has both LTR/RTL languages, I think it would be desirable to maintain the physical position in either language type.

**Comment by t-hamano** [comment]
Thanks for the approval!
I would like to ask someone to review the behavior on the cover block before merging.

@draganescu
@aaronrobertshaw

If I may, I would appreciate a review of this PR.
You seemed to have worked on the cover block.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for working through this one @t-hamano 👍 

From what I can gather this appears to be testing as advertised.

✅ Grid is maintained and tooltips match in Storybook regardless of language direction
✅ On trunk the tooltips didn't match the physical direction of the grid
✅ With this PR's changes applied, the tooltips match the physical direction of the grid
✅ Order of the blocks appear consistent before and after these changes

❓While testing I did notice that the block toolbar is out of position. I'm guessing this isn't related to this PR but it wasn't an issue on trunk. Perhaps just a rebase?

It might also help others coming to this PR with limited history or context if we can get some before/after demos in the PR description. 

| Trunk - RTL | This PR - RTL |
|---|---|
| Screen Shot 2022-08-22 at 5 38 42 pm | Screen Shot 2022-08-22 at 5 42 19 pm |

**Comment by t-hamano** [comment]
> While testing I did notice that the block toolbar is out of position. I'm guessing this isn't related to this PR but it wasn't an issue on trunk. Perhaps just a rebase?

I have looked into this and it appears that this is not a problem that is limited to this PR and cover block.
Presumably, if no blocks are selected and a new block is selected, the block toolbar will shift.

https://user-images.githubusercontent.com/54422211/185952724-6a3ae88e-a316-4bd0-a270-3154f6f9e311.mp4

Perhaps this should be addressed as a separate issue.

> It might also help others coming to this PR with limited history or context if we can get some before/after demos in the PR description.

I understand, and I will keep that in mind from now on 🙂

File: packages/data/tsconfig.json [hunk 2]
```diff
@@ -14,7 +14,5 @@
 		{ "path": "../priority-queue" },
 		{ "path": "../redux-routine" }
 	],
-	"include": [
-		"src/**/*"
-	]
+	"include": [ "src/**/*" ]
```
Comment (t-hamano) [inline packages/data/tsconfig.json:17]: When the latest trunk was merged, unexpected code reformatting was performed.
But since this change was made by the linter, it appears to be OK.

**Comment by t-hamano** [comment]
@aaronrobertshaw 
I have created a new issue #43605 for the problem you mentioned in [this comment](https://github.com/WordPress/gutenberg/pull/43126#pullrequestreview-1080018363).
I would like to merge it as soon as the test passes, as I think this PR isn't the cause👍

**Comment by aaronrobertshaw** [comment]
> I have created a new issue https://github.com/WordPress/gutenberg/issues/43605 for the problem you mentioned in https://github.com/WordPress/gutenberg/pull/43126#pullrequestreview-1080018363.

Thanks for creating the issue and merging this. 👍
